src/test/common/test_async_yield_waiter.cc

```diff
@@ -240,4 +242,28 @@ TEST(YieldWaiterPtr, wait_error)
   }
 }
 
+void invoke_callback(int expected_reply, std::function<void(int)> cb) {
+  auto t = std::thread([cb, expected_reply] {
+      std::this_thread::sleep_for(std::chrono::seconds(1));
```

Comment:
i've really been trying to avoid sleeps in unit tests. short sleeps make for flaky tests, and long sleeps make them slow (that adds up in 'make check'). if the test cases are fast, you can run them with something like `--gtest_repeat=1000` under valgrind or sanitizers to help catch any races

in this case, you just need to ensure that `cb()` is called after `waiter.async_wait()` suspends

`boost::asio::defer()` is a useful tool for this. it schedules a handler to run after the current one finishes. you can use that to avoid the sleep by spawning the thread in a deferred block:

```diff
-        invoke_callback(expected_reply, [&waiter](int r) {waiter.complete(boost::system::error_code{}, r);});
+        // trigger completion after async_wait() suspends
+        boost::asio::defer(yield.get_executor(), [&] {
+            invoke_callback(expected_reply, [&waiter](int r) {waiter.complete(boost::system::error_code{}, r);});
+          });
         reply = waiter.async_wait(yield);
```

`cross_thread_cancel` is a recent example that does something similar, although it spawns a separate thread for the execution context, waits on a latch for that thread to block in throttle.wait(), then triggers cancellation from the main thread: https://github.com/ceph/ceph/pull/58348/commits/ac4c8adb02683c168fedafdae64e379050a86245

lack of sleeps there keep the runtime extremely short:
> 12 tests from co_throttle (3 ms total)

Reply:
hmm. calling "defer" (with or without the timeout) fixes the tsan errors.
maybe this is the root cause for the problem?
will try to modify that in `rgw_pubsub_push.cc`

Reply:
changed the notifications code to use defer